packages/base-controller/src/BaseControllerV2.ts

```diff
@@ -250,29 +258,30 @@ function deriveStateFromMetadata<S extends Record<string, Json>>(
   metadata: StateMetadata<S>,
   metadataProperty: 'anonymous' | 'persist',
 ): Record<string, Json> {
-  return Object.keys(state).reduce((persistedState, key) => {
-    try {
-      const stateMetadata = metadata[key as keyof S];
-      if (!stateMetadata) {
-        throw new Error(`No metadata found for '${key}'`);
+  return Object.keys(state).reduce(
+    (persistedState: Record<string, Json>, key: keyof S) => {
+      try {
+        const stateMetadata = metadata[key];
+        if (!stateMetadata) {
+          throw new Error(`No metadata found for '${String(key)}'`);
+        }
+        const propertyMetadata = stateMetadata[metadataProperty];
+        const stateProperty = state[key];
+        if (typeof propertyMetadata === 'function') {
+          persistedState[String(key)] = propertyMetadata(stateProperty);
+        } else if (propertyMetadata) {
+          persistedState[String(key)] = stateProperty;
```

Comment:
Why all the `String`s?

Reply:
Figured out a better way to handle this: b32a7ed537591e8fa0a63786d09f327e0a632a7f

Reply:
This is fundamentally because `keyof Record<string, ...>` is `string | number | symbol`.

Reply:
Further updated here: ee2467dafecd8ca83c736589ff9462e74421ee38
See: https://github.com/MetaMask/core/pull/1890#discussion_r1390574901